sqladmin/application.py

```diff
@@ -475,7 +475,8 @@ async def edit(self, request: Request) -> Response:
                 status_code=400,
             )
 
-        await model_view.update_model(pk=request.path_params["pk"], data=form.data)
+        form_ = model_view.on_model_change(data=form.data, pk=request.path_params["pk"])
+        await model_view.update_model(pk=request.path_params["pk"], data=form_)
```

Comment:
And instead of changes in `application.py` I think you can add them in `_queries.py` so it's running in the same transaction:

```py
    def _update_sync(self, pk: Any, data: Dict[str, Any]) -> None:
        pk = get_column_python_type(self.model_view.pk_column)(pk)
        stmt = select(self.model_view.model).where(self.model_view.pk_column == pk)

        with self.model_view.sessionmaker() as session:
            obj = session.execute(stmt).scalars().first()
            anyio.from_thread.run(self.model_view.on_model_change, data, obj, False)
            obj = self._set_attributes_sync(session, obj, data)
            session.commit()
            anyio.from_thread.run(self.model_view.after_model_change, data, obj, False)

    async def _update_async(self, pk: Any, data: Dict[str, Any]) -> None:
        pk = get_column_python_type(self.model_view.pk_column)(pk)
        stmt = select(self.model_view.model).where(self.model_view.pk_column == pk)

        for relation in self.model_view._relations:
            stmt = stmt.options(joinedload(relation.key))

        async with self.model_view.sessionmaker() as session:
            result = await session.execute(stmt)
            obj = result.scalars().first()
            await self.model_view.on_model_change(data, obj, False)
            obj = await self._set_attributes_async(session, obj, data)
            await session.commit()
            await self.model_view.after_model_change(data, obj, False)
```

Reply:
You can do the same for insert method.